### ISSUE TITLE
Automatically exclude Play routes files from formatting

### ISSUE THREAD

**Issue by balagez** [sbt/sbt-scalariform#32]
I noticed that when using Play 2.4.x with sbt-scalariform incremental compilation breaks because sbt-scalariform is constantly reformatting the generated routes files.

I was able to overcome this by manually excluding those files from formatting:

``` scala
excludeFilter in scalariformFormat := (excludeFilter in scalariformFormat).value ||
    "Routes.scala" ||
    "ReverseRoutes.scala" ||
    "JavaScriptReverseRoutes.scala" ||
    "RoutesPrefix.scala"
```

This is not the best solution because if you have another file called `Routes.scala` somewhere else in your code if won't be formatted. It would be much better if they worked more smoothly with each other out of the box.

Also note that the documentation says `excludeFilter in format` which is no longer valid and doesn't compile so please change the relevant parts of README to `excludeFilter in scalariformFormat`.

**Comment by fommil** [comment]
there is documentation in the README about how to exclude code blocks using comments. Reopen if that does not address your use case.

A PR to the README to change the parts that you feel are out of date would be much appreciated.

**Comment by fommil** [comment]
this one => https://github.com/scala-ide/scalariform

**Comment by balagez** [comment]
Thanks, it does not address my case because I'm not a maintainer of Play Framework but I'll create PR for them.

And here's the follow-up PR for the documentation: https://github.com/sbt/sbt-scalariform/pull/33

**Comment by fommil** [comment]
you put comments in your code to stop your code from being reformatted. Why do you need to be a maintainer of Play for that?

**Comment by balagez** [comment]
Because it's the routes files which are generated from the source routes files by Play, i.e. `Routes.scala`, `ReverseRoutes.scala`, etc. in `crossTarget`, that prevent the incremental compiler from working correctly when Scalariform is set up.

In Play 2.4.x those `.scala` files are no longer in `srcManaged` but in `crossTarget / "{routes file name}" / "main" / "router"` or something similar, so Scalariform finds and reformats them according to the current formatting rules, which might be different from what the Play guys used in their [routes templates](https://github.com/playframework/playframework/tree/master/framework/src/routes-compiler/src/main/twirl/play/routes/compiler/static).

As a user of Play, I have no control over what code the framework generates.

**Comment by fommil** [comment]
ok, I see. I think this is perhaps a question for play rather than sbt scalariform. If file path/name matching isn't enough there isn't much we can do here.

**Comment by jroper** [comment]
There is a lot you can do actually, sbt makes this distinction really easy, scalariform should only be run on `unmanagedSources` (and `unmanagedSourceDirectories`), source generators generate sources into `managedSources` (and `managedSourceDirectories`).  Having said that, when I look at the source code for sbt-scalariform, it seems to be finding its sources by looking at `scalaSource`, which is generally the `src/main/scala` directory, or in Play, it's the `app` directory.  This will never contain generated sources, so the current code I don't think exhibits this bug.  Maybe the reporter is using an old version of scalariform that did exhibit this bug, or maybe they have a configuration that is adding all `sourceDirectories in Compile` to the scaliform `sourceDirectories`.  Or, I've misread the scalaiform code, and it is using `sourceDirectories in Compile` instead of `unmanagedSourceDirectories in Compile`.

**Comment by balagez** [comment]
A little more detail here: I used Play Framework 2.4.6 with sbt-scalariform 1.6.0, I guess both are the latest release versions.

Reproducing is as simple as cloning the play-scala activator template, adding sbt-scalariform to project/plugins.sbt and then running sbt compile a few times without changing any lines.

Console output below. At the end I printed the values of `scalaSource`, `unmanagedSourceDirectories`, `sourceDirectories`, `managedSources` and `playRoutes`.

```
~ $ activator new myproject play-scala

Fetching the latest list of templates...

OK, application "myproject" is being created using the "play-scala" template.

To run "myproject" from the command line, "cd myproject" then:
~/myproject/activator run

To run the test for "myproject" from the command line, "cd myproject" then:
~/myproject/activator test

To run the Activator UI for "myproject" from the command line, "cd myproject" then:
~/myproject/activator ui

~ $ cd myproject/
~/myproject $ echo -e '\naddSbtPlugin("org.scalariform" % "sbt-scalariform" % "1.6.0")' >> project/plugins.sbt
~/myproject $ sbt
[info] Loading global plugins from ~/.sbt/0.13/plugins
[info] Loading project definition from ~/myproject/project
[info] Updating {file:~/myproject/project/}myproject-build...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] Set current project to myproject (in build file:~/myproject/)
[myproject] $ compile
[info] Updating {file:~/myproject/}root...
[info] Resolving org.apache.commons#commons-lang3;3.4 ...
[info] Formatting 1 Scala source {file:~/myproject/}root(compile) ...
[info] Resolving jline#jline;2.12.1 ...
[info] Done updating.
[info] Compiling 7 Scala sources and 1 Java source to ~/myproject/target/scala-2.11/classes...
[success] Total time: 9 s, completed Jan 27, 2016 7:54:49 AM
[myproject] $ compile
[info] Formatting 6 Scala sources {file:~/myproject/}root(compile) ...
[info] Reformatted 6 Scala sources {file:~/myproject/}root(compile).
[info] Compiling 7 Scala sources and 1 Java source to ~/myproject/target/scala-2.11/classes...
[success] Total time: 3 s, completed Jan 27, 2016 7:54:54 AM
[myproject] $ compile
[info] Formatting 4 Scala sources {file:~/myproject/}root(compile) ...
[info] Reformatted 4 Scala sources {file:~/myproject/}root(compile).
[info] Compiling 7 Scala sources and 1 Java source to ~/myproject/target/scala-2.11/classes...
[success] Total time: 2 s, completed Jan 27, 2016 7:54:57 AM
[myproject] $ compile
[info] Formatting 4 Scala sources {file:~/myproject/}root(compile) ...
[info] Reformatted 4 Scala sources {file:~/myproject/}root(compile).
[info] Compiling 7 Scala sources and 1 Java source to ~/myproject/target/scala-2.11/classes...
[success] Total time: 2 s, completed Jan 27, 2016 7:55:03 AM
[myproject] $ 
[myproject] $ scalaSource
[info] ~/myproject/app
[myproject] $ unmanagedSourceDirectories
[info] List(~/myproject/app-2.11, ~/myproject/app, ~/myproject/app)
[myproject] $ sourceDirectories
[info] List(~/myproject/app-2.11, ~/myproject/app, ~/myproject/app, ~/myproject/target/scala-2.11/src_managed/main, ~/myproject/target/scala-2.11/twirl/main, ~/myproject/target/scala-2.11/routes/main)
[myproject] $ show managedSources
[info] List(~/myproject/target/scala-2.11/twirl/main/views/html/index.template.scala, ~/myproject/target/scala-2.11/twirl/main/views/html/main.template.scala, ~/myproject/target/scala-2.11/routes/main/controllers/routes.java, ~/myproject/target/scala-2.11/routes/main/router/Routes.scala, ~/myproject/target/scala-2.11/routes/main/controllers/ReverseRoutes.scala, ~/myproject/target/scala-2.11/routes/main/controllers/javascript/JavaScriptReverseRoutes.scala, ~/myproject/target/scala-2.11/routes/main/router/RoutesPrefix.scala)
[success] Total time: 0 s, completed Jan 27, 2016 7:55:18 AM
[myproject] $ show playRoutes
[info] Vector(~/myproject/target/scala-2.11/routes/main/controllers/routes.java, ~/myproject/target/scala-2.11/routes/main/router/Routes.scala, ~/myproject/target/scala-2.11/routes/main/controllers/ReverseRoutes.scala, ~/myproject/target/scala-2.11/routes/main/controllers/javascript/JavaScriptReverseRoutes.scala, ~/myproject/target/scala-2.11/routes/main/router/RoutesPrefix.scala)
[success] Total time: 0 s, completed Jan 27, 2016 7:55:47 AM
```

**Comment by balagez** [comment]
I don't think I can reopen this issue but as the example above shows it does cause problems for all users wanting to use sbt-scalariform in conjunction with Play Framework, I think this is something that should be addressed.

**Comment by jroper** [comment]
@fommil Do you know where 1.6.0 was cut from?  There's no tag for me to look at the source code.

**Comment by sschaef** [comment]
1.6.0 release was built by @daniel-trinh. He didn't publish the tags so far.

**Comment by fommil** [comment]
@jroper found the cause of this, but we were disrupted by the github outage last night

**Comment by jroper** [comment]
The problem I believe is caad9a95086136e40a6b82681d381dc4a8c45a15, and the specific line is:

```scala
(sourceDirectories in Global in scalariformFormat) := List(scalaSource.value),
```

By setting `compile:scalariformFormat::sourceDirectories` in global scope, when a project goes to use it later, and looks up `someproject/compile:scalariformFormat::sourceDirectories`, because of the way sbt delegates work, it falls back to `someproject/compile:sourceDirectories` before it falls back to  `*/compile:scalariformFormat::sourceDirectories`, and so finds the regular compile source directories which is both managed and unmanaged source directories.

But, setting it at global scope doesn't make any sense at all, the effect of this setting, when applied to a multiproject build, is that the last project to have its settings processed will be the one that set `sourceDirectories` in global - so if we made scalariform read this setting, for each project that it was run on, it would run on the sources of that last project that set it.  And, it wouldn't make sense to try and configure this globally (eg in the `~/.sbt` folder), because globally, you don't even have a `baseDirectory` configured to be able to get a source directory from, since `baseDirectory` is only configured in `ThisBuild`, not globally (because there is no global base directory, builds have base directories).

So unless @jrudolph had a good reason for making `sourceDirectories` configurable globally, I think that line should be reverted.

**Comment by jrudolph** [comment]
I agree with @jroper. That line should be reverted. I guess the idea was to put the setting into some lower priority "default" scope but, of course, it shouldn't change anything outside of this projects scope.

**Comment by JMarotzke** [comment]
Hi, to me it seems as if the pull request #43 solves this issue. Will this ever be merged and released?

**Comment by gmalouf** [comment]
It would be great if one of the pull requests #42 or #43 were merged in..

**Comment by justjoheinz** [comment]
Please fix this :-)

**Comment by mkokho** [comment]
Ah, fix #43 is in not 1.7.0 release...
Well, here is a workaround. Add this to your `build.sbt`
```
sourceDirectories in (Compile, scalariformFormat) := (unmanagedSourceDirectories in Compile).value
```

If you want to debug settings, try `inspect` and `show` commands in the sbt console. They will show task dependencies, their scopes and axis 
```
[root] $ compile:scalariformFormat::sourceDirectories
[root] $ inspect scalariformFormat
[root] $ show show scalariformPreferences
```

**Comment by godenji** [comment]
@mkokho will be in next release.

**Comment by henricook** [comment]
Sorry for the noob question, i'm attempting to implement the workaround but `scalariformFormat` isn't a thing in my build.sbt - where does it come from?

**Comment by godenji** [comment]
@henricook the base imports are:

```
import scalariform.formatter.preferences._
import com.typesafe.sbt.SbtScalariform
import com.typesafe.sbt.SbtScalariform.ScalariformKeys
```

`scalariformFormat` sbt task lives in `SbtScalariform`, so you could import it or just do `SbtScalariform.scalariformFormat := ...`.

1.7.1, which fixes this issue, was just released -- should be up on Maven Central in a couple of hours.

**Comment by godenji** [comment]
fixed in 1.7.1

**Comment by balagez** [comment]
It was indeed fixed in 1.7.1 but now it's broken again in 1.8.0
